|          | H2 E.coli 1/2 | B9 E. coli 1/2 | H3 E.c+M.a 1/2 | E. coli | M.aerug. | R.slithy |
|:---------|--------------:|---------------:|---------------:|--------:|---------:|---------:|
|seq1097   |      1129/1129|         417/319|           40/31|        7|        0 |        0 |
|seq595    |            1/0|             0/0|             0/0|        0|        0 |        0 |
|seq8683   |            1/0|             1/0|             0/0|        0|        0 |        0 |
|seq10871  |            0/0|             0/0|             1/0|        0|        0 |        0 |
|seq120556 |            0/0|             0/0|             1/2|        0|        0 |        2 |
|seq1892   |            0/0|             0/0|             3/3|        0|        0 |        0 |
|seq51132  |            0/0|             0/0|             1/0|        0|        0 |        0 |
|seq953    |            0/0|             0/0|             9/4|        0|        0 |        0 |
|seq62     |            0/0|             0/0|             0/1|        0|        0 |        0 |
|seq87355  |            0/0|             0/0|             0/1|        0|        0 |        0 |
|seq421    |            0/0|             0/0|             0/0|        0|        2 |        0 |

- NC_012892.2 Escherichia coli BL21(DE3)
- NC_010296.1 Microcystis aeruginosa
- NC_015703.1 Runella slithyformis DSM 19594

>seq595_k__Bacteria;p__Bacteroidetes;c__Sphingobacteria;o__Sphingobacteriales;f__Chitinophagaceae;g__;s__
>seq8683_k__Bacteria;p__Proteobacteria;c__Gammaproteobacteria;o__Enterobacteriales;f__Enterobacteriaceae;g__;s__
>seq10871_k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Erythrobacteraceae;g__Porphyrobacter;s__
>seq120556_k__Bacteria;p__Bacteroidetes;c__Sphingobacteria;o__Sphingobacteriales;f__Cytophagaceae;g__Runella;s__
>seq1892_k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Novosphingobium;s__
>seq51132_k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Sphingomonadales;f__Sphingomonadaceae;g__Sphingobium;s__
>seq953_Bacteria_Cyanobacteria_Oscillatoriophycideae_Chroococcales_Microcystaceae_Microcystis__MicrocystisaeruginosaNIES-843
>seq62_k__Bacteria;p__Cyanobacteria;c__Cyanobacteria;o__;f__Chloroplast;g__Bacillariophyta;s__
>seq87355_k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rhodospirillales;f__;g__;s__

To assign the bins to a 16S sequence, we can perform supervised & unsupervised classification. The former can be implemented as a random forest and the latter as heirarchical cluster analysis. Both analyses were performed on mixed matrices including the full hierarhy of taxanomic classifications and numeric vectors corresponding to abundances in the water column. The script that performed the analysis is called `classify_and_cluster.R` and it is provided the test data set, train data set, and bin data set. The files corresponding to these inputs (provided in order on the command line) are called `otu_abund_tax_train.tsv`, `smg_abund_tax_test.tsv`, and `bin_abund_taxa_mat.tsv`. 

The RF model was trained using the full data set of OTU abundances across all depths and times and tested on the amplicons recovered directly from alignment with the shotgun metagenomic libraries, as these labels were known. 

Inter-amplicon distances were calculated using the `daisy()` function from the `cluster` package, which employs Gower's distance as the default output. Gower's method is akin to euclidean distance, but was selected as it is superior for use on mixed data types. The R command `hclust` was used to perform hierachical clustering using Ward's method. The object returned by `hclust` contains a `merge` vector which shows the order in which samples & clusters are agglomerated. It also contains a `height` vector containing a method specific numerical measure of the distance bridged at each step to create a new grouping. The algorithm proceeds by merging pairs of samples producing smaller `height` values at first, with greater `height` values recorded at every step. 

http://www.sthda.com/english/articles/26-clustering-basics/86-clustering-distance-measures-essentials/
http://www.r-tutor.com/gpu-computing/clustering/hierarchical-cluster-analysis


hclust(d, "ward.D2"